[AOTInductor] Initial functionality for Inf and NaN checker

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109526

Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler @ngimel

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

muchulee8 committed:
[AOTInductor] Initial functionality for Inf and NaN checker

Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49379751).

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

mlazos commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1018,6 +1023,58 @@ def write_header(self):
+              }
+
+              if (assertion) {
+                assert(!has_nan);
```
I think it would be useful to report the offending indices if possible (not sure what's possible in our C++ infra)

muchulee8 replied:
I added a TODO at [line 1515](https://github.com/pytorch/pytorch/pull/109526/files#diff-18714c66f0b1f90de1bfd784bae5fab68a4e54d817556fefa5cdee2621cb06d0R1515). We are still settling with how log would look like (likely c10 logging would be removed)

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1018,6 +1023,58 @@ def write_header(self):
+              }
+
+              if (assertion) {
+                assert(!has_inf);
```
Is this `assert` macro from `<assert.h>`? Will it be optimized away with an `-O` flag in the compilation command?

muchulee8 replied:
No, it won't be optimized away. In the test being ran locally, it do assert.

aakhundov commented on test/inductor/test_aot_inductor.py:
```diff
@@ -176,20 +179,51 @@ def forward(self, x, y):
+            example_inputs,
+            expected,
+            options={"debug_check_inf_and_nan": True},
+        )
```
Should we assert here?

muchulee8 replied:
hmmm.. it doesn't really matter.
Line 207 doesn't really run, because the assert in C++ can't be captured by python code. It will just fail out directly. Also, even if we have some mechanism to "continue", I don't think comparison of NaN is meaningful in Python (i.e. `float("nan") != float("nan"`)

mlazos replied:
You could use torch.isnan to do this comparison (call it on each tensor and check the results match) not sure if this is worth it if things fail out before.

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49379751).

muchulee8 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1452,6 +1509,12 @@ def generate_profiler_mark_wrapper_call(self, stack):
         )
 
+    def generate_inf_and_nan_checker(self, nodes):
+        self.write_inf_and_nan_header()
```
guarded in line 1027.

chenyang78 reviewed (approved):
LGTM. Thanks!

muchulee8 commented:
@pytorchbot merge

muchulee8 commented:
@pytorchbot label "topic: not user facing"

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng aakhundov

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler ngimel

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler ngimel

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Initial functionality for Inf and NaN checker"


Summary:
Add initial functionality for Inf and NaN checker for AOTInductor.

Test Plan:
Included in commit. Skipped for CI as SIGABRT can't be captured by pytest.

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler ngimel

Differential Revision: [D49379751](https://our.internmc.facebook.com/intern/diff/D49379751)

[ghstack-poisoned]

khabinov commented on test/inductor/test_aot_inductor.py:
```diff
@@ -394,6 +394,50 @@ def forward(self, x):
         example_inputs = (torch.rand(6, device=self.device),)
         self.check_model(Model(), example_inputs)
 
+    @unittest.skip("Skip this test, only for local test. SIGABRT is produced.")
```
Can we use `signal.signal()` to catch `SIGABRT`? https://docs.python.org/3/library/signal.html#signal.signal

muchulee8 replied:
Unfortunately no :( 
It's still a core-dump, I tried to rehook `SIGABRT` but still doesn't work. I don't know the reason, but from some stackoverflow that I read, it seems like I need to hook it with ctypes at the very moment of exec (which, in our case seems to be `AOTModelRunner`) and that would be a huge change and non-trivial with our current testing setup. This isn't a test that must-not break, since it's mostly a debugging feature, so it should be fine.

khabinov commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -301,6 +301,21 @@ AOTITorchError aoti_torch_mm_out(
+    for (int64_t i = 0; i < flattened.numel(); i++) {
+      auto value = flattened[i].item<float>();
+      if (std::isinf(value) || std::isnan(value)) {
+        assert(false);
```
Are we ok with using `assert` here? It will be wiped out without `NDEBUG`.

muchulee8 replied:
I think it should be fine. We will not have this in prod or for testing etc. This, in theory, would only be plugged in when we are debugging.

muchulee8 commented:
@pytorchbot merge
